New blog post (#598)
ePages-de/epages-devportal: _posts/2020/2020-02-11-unit-testing-in-react.md

---
layout: post
title: Unit testing in React
date: 2020-02-11
header_image: public/code-coding-computer-cyberspace-270373.jpg
header_overlay: true
category: coding
tags: ["javascript", "reactjs", "testing"]
authors: ["Daniel"]
about_authors: ["dhara"]
---

Automated Testing is one of the most important aspects of software development. You can test your code in different ways and levels, the most relevant
of which are unit testing and integration testing.
On the one hand, in unit testing, you test each one of your functions and components individually. On the other, in integration testing, you test
how all those work along with each other to deliver all the fancy features you offer your users.

> As Adrià Fontcuberta pointed out at his remarkable talk at dotJS in Paris (check out this [blog post](/blog/events/dotjs-2019-in-paris-from-the-perspective-of-a-frontend-designer/){:target="_blank"} about the conference, as well as the [video](https://www.dotconferences.com/2019/12/adria-fontcuberta-the-pragmatic-front-end-tester){:target="_blank"} of the actual talk), tests give predictability and

Reviewer: I would recommend to add "conference" after "dotJS".

"help us sleep well at night".

In this blog post, I'll show you useful hints for unit testing a simple component in React.

## Getting started

> One of our most recent features at our storefront was a "Load more" button in the product pages. We could draft here a simplified version of this

Reviewer: I think it must be "... on the product pages".

> One of our most recent features at our storefront was a "Load more" button in the product pages. We could draft here a simplified version of this

Author: done

> One of our most recent features at our storefront was a "Load more" button in the product pages. We could draft here a simplified version of this

Author: done

> feature, and carry on to writing tests for it! This will help us to understand how tests work and how to ensure our tests are making our code

Reviewer: Sugesstion: "... and how to ensure that ..."

Author: In this case, I guess it is indifferent. I find the variant without "that" more direct and closer to spoken informal dialogue.

reliable and predictable.

> We could start out with a simple component which has a button which can fetch (via an API) a fixed amount of, say, 1 product, and stores them

Reviewer: Suggestion to avoid "which" twice in a sentence: "... a button that ..."

Author: > Is "1 product" an example? If so, it's maybe better to write for example: "e.g."
Yes, the use of the word "say" already highlights that the number of "1" is an arbitrary suggestion.

in its state in order to subsequently display it. Our first shot could be:

```javascript
import React, { useState } from 'react';

const Product = (product) => (
  <div>
    <div>{product.id}</div>
    <div>{product.name}</div>
    <div>{product.price}</div>
  </div>

> )

Author: done

const fetchProducts = (page) => fetch(`http://localhost/products?page=${page}`)
  .then(response => response.json())

>   .catch((error) => console.log(error))

Author: done

const ProductList = () => {
  const [products, setProducts] = useState([])
  const [page, setPage] = useState(1)
  const loadProducts = async () => {
    const response = await fetchProducts(page)
    if(response) {
      const newProducts = response.products
      setProducts(products.concat(newProducts))
      setPage(page + 1)
    }
  }
  return <div>
    <button onClick={loadProducts}>
       Load more!
    </button>
    {products.map(product => <Product {...product} />)}

>   </div>

Author: Using ```jsx fixes it!

}

export default ProductList;
```

> Before going on to how to test it, I must say I've prepared a sample repo for you to try out all the code in this post.

Reviewer: Maybe you can write the exact word, instead of "it": "Before going on to how to test **it**, ..."

Author: I didn't understand this comment. What do you mean?

> Before going on to how to test it, I must say I've prepared a sample repo for you to try out all the code in this post.

Author: done

Just check it out on Github [here](https://github.com/DanielHara/unit-testing-react){:target="_blank"}.

## Let's test it!

> Now comes the question, how would you test the component `ProductList`?

Reviewer: Suggestion: " Now we come to the question, ..."

Author: Why? It's the same thing.

> Here at ePages we use extensively the libraries `react-testing-library` and `jest` to write our unit tests.

Author: done

The former contains lots of useful features, like querying for HTML elements and firing all the events the user can trigger, while the
latter provides us with a thorough mocking and assertion library.
As soon as I'm done writing the first lines of the tests, I imagine that my first test case would be to try to find a button
to load some products:

```javascript
import { fireEvent, render } from '@testing-library/react'

it('should render a button to load products', () => {
  const { getByText } = render(<ProductList />)

  const loadMoreButton = getByText('Load more!')
  fireEvent.click(loadMoreButton)
})
```

But, now, what I should expect? I do not know what some external API will return me. Maybe it is temporarily down.
There comes an important notion of testing: you must decouple your tests from other dependencies as much as possible. After all, are you testing
just `ProductList` or also an external API?

## Mock it

`jest` offers you the possibility of mocking all kinds of functions, including their return values and assert how many times they were called,

> and with which arguments they were called. It also offers a great deal of synthatic sugar to make your tests look pretty and for you to admire them

Author: done

after they've been written.

Maybe we could mock `fetchProducts` to take a look on whether it was called with the right arguments, and also mock the return value,
to assert whether the `ProductList` is rendering them correctly. This way our tests becomes a lot more predictable.

However... Can you do it? It's now difficult to mock `fetchProducts`, because you'd have to override the `ProductList.jsx` file.
There comes another hint: try to inject your external dependencies such that you can mock them! You could pass `fetchProducts` as a prop instead:

```javascript
const defaultFetchProducts = (page) => fetch(`http://localhost/products?page=${page}`)
  .then(response => response.json())

>   .catch((error) => console.log(error))

Reviewer 2: For brevity in this blog post, I suggest to leave out the `.catch()`.

Author: done

> const ProductList = ({ fetchProducts=defaultFetchProducts }) => {

Reviewer 2: ```suggestion
const ProductList = ({ fetchProducts = defaultFetchProducts }) => {
```

Author: done - I linted it quickly with standard

  const [products, setProducts] = useState([])
  const [page, setPage] = useState(1)
  const loadProducts = async () => {
    const response = await fetchProducts(page)

>     if(response) {

Reviewer 2: ```suggestion

    if (response) {
```

Author: done

      const newProducts = response.products
      setProducts(products.concat(newProducts))
      setPage(page + 1)
    }
  }
  return <div>
    <button onClick={loadProducts}>
       Load more!
    </button>
    {products.map(product => <Product {...product} />)}
  </div>
}
```

And what about the test?

```javascript
import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import ProductList from '../ProductList'
import Bluebird from 'bluebird'
import { act } from 'react-dom/test-utils'

it('should render a button to load products', async () => {
  const fetchProducts = jest.fn().mockImplementation((page) => Promise.resolve({
    products: [{
      id: `dummy-id-${page}`,
      name: 'dummy-product',
      price: '10 £',
    }]
  }))

  const { getByText } = render(<ProductList fetchProducts={fetchProducts}/>)

  const loadMoreButton = getByText('Load more!')

  // click the first time
  fireEvent.click(loadMoreButton)
  await act(() => Bluebird.delay(500))

  // expect one product
  expect(getByText('dummy-id-1')).toBeTruthy()
  expect(fetchProducts).toHaveBeenCalledTimes(1)
  expect(fetchProducts).toHaveBeenCalledWith(1)

>   //click the second time

Reviewer 2: For consistency:

```suggestion
  // click the second time
```

Author: done

  fireEvent.click(loadMoreButton)
  await act(() => Bluebird.delay(500))

  // expect two products
  expect(getByText('dummy-id-1')).toBeTruthy()
  expect(fetchProducts).toHaveBeenCalledTimes(2)
  expect(fetchProducts).toHaveBeenCalledWith(2)
})
```

This test can already be taken seriously! It checks whether we called `fetchProducts` with the right arguments and whether we
use the result of these calls in a meaningful way.

> On the repo, this approach is on branch `mocking-fetch-products`.

Reviewer 2: I would avoid the short notation "repo" in favour of writing the full word.

```suggestion
In the example repository, you can find this approach on the branch `mocking-fetch-products`.
```

Author: done

## Taking it to the next level

The former approach has, nevertheless, its drawbacks. We rely completely on the mock of `fetchProducts`. How can we know if it would
hit the right API endpoints?

> There's were the awesome `nock` library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will

Reviewer 2: Same as written in the other annotation: I think we shouldn't use the package names, but instead the official spelling of the names according to the corresponding libraries (and possibly link them for the first occurrence in the text).

```suggestion
That's were the awesome Nock library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will
```

Author: done - but here I really mean "There"

> mock any requests made to `http://localhost` _inside_ our test! This way we also test the right HTTP requests are being made

Reviewer 2: ```suggestion
mock any requests made to `http://localhost` _inside_ our test! This way, we also test that the right HTTP requests are being made
```

Author: done

and do not have to mock `fetchProducts` any all!

```javascript
import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import nock from 'nock'
import ProductList from '../ProductList'
import Bluebird from 'bluebird'
import { act } from 'react-dom/test-utils'

it('should render a button to load products', async () => {
  const scope = nock(`http://localhost`)
    .get('/products')
    .query({ page: 1 })
    .reply(200, {
      products: [{
        id: `dummy-id-1`,
        name: 'dummy-product-1',
        price: '10 £',
      }]
    })
    .get('/products')
    .query({ page: 2 })
    .reply(200, {
      products: [{
        id: `dummy-id-2`,
        name: 'dummy-product-2',
        price: '20 £',
      }]
    })

  const { getByText } = render(<ProductList />)

  const loadMoreButton = getByText('Load more!')

  fireEvent.click(loadMoreButton)
  await act(() => Bluebird.delay(500))

  // expect one product
  expect(getByText('dummy-id-1')).toBeTruthy()

  fireEvent.click(loadMoreButton)
  await act(() => Bluebird.delay(500))

>   //expect two products

Reviewer 2: ```suggestion
  // expect two products
```

Author: done

  expect(getByText('dummy-id-1')).toBeTruthy()
  expect(getByText('dummy-id-2')).toBeTruthy()

  scope.done()
})
```

> You can try this out on the `master` branch of the repo!

Reviewer 2: _See the other annotation_.

```suggestion
You can try this out on the `master` branch of the example repository.
```

Also, I think these hint paragraphs to the example repository would look good in a info box, too (see other annotation).

Author: done

Now, we actually test that the endpoint `/products` has been hit with query params `page=1` and `page=2`, thanks to the
`scope.done()` call, which is an awesome feature of `nock`! It asserts that all of the mocked API endpoint calls have been hit!
And, furthermore, this tests the component's feature independently of its implementation (so called 'black-box testing').
It does not matter for the test, for instance, how the `page` parameter is stored in the component. This makes it a lot easier

> when the need for _refactoring_ comes.

Reviewer 2: ```suggestion
when the component needs to be refactored at a later point.
```

Author: done

Reviewer 2: But, it's the same as before. Is this intended?

## Conclusion

Nowadays, there is absolutely no reason why you should not unit test your code. With so many great tools out there, the effort

> of writing tests absolutely pays off by dramatically decreasing the quantity of bugs and getting you covered when refactoring.  🎉

Reviewer 2: 👍
